### ISSUE TITLE
StrategyKeeperDAOGeneric

### ISSUE THREAD

**Issue by hsukev** [yearn/yearn-strategies#47]
**Description**: 
Strategy earns rewards from KeeperDAO pool and redeposits earnings back into same pool.

Complication comes from KeeperDAO's 0.64% deposit fee. This strategy keeps track of that internally as `incurredLoss` and optimizes `harvestTrigger` to when `profit > incurredLoss`

`want` can be any of the tokens supported by the KeeperDAO pool (DAI, USDC, renBTC, WETH) except for ETH (need some refactoring for ETH)

reward claiming requires off-chain data from 
https://indibo-lpq2.herokuapp.com/reward_of_liquidity_provider/YOUR_ADDRESS
https://indibo-lp.herokuapp.com/reward_of_liquidity_provider/YOUR_ADDRESS

**Commit (hash)**: 

**Tag**:

**Repo**: https://github.com/hsukev/KeeperDAO

**Scope**: (List link to files)
- [Strategy.sol](https://github.com/hsukev/KeeperDAO/blob/main/contracts/Strategy.sol)

**Additional References or Repo** (Link to underlying protocols or relevant documents):

**Deployed Contract** (etherscan link): 
not deployed

**Review Ongoing By**:
- [ ] Dev 1
- [ ] Dev 2

**Review Completed By**:
- [ ] Dev 1
- [ ] Dev 2

**Comment by matkam** [comment]
Is there any other way to test this? This is probably not a good idea to leave publicly accessible
https://github.com/hsukev/KeeperDAO/blob/a2b65aee342e5217678e9a3a8649ba5f82dfa9a8/contracts/Strategy.sol#L90

**Comment by matkam** [comment]
This doesn't need to be `public`, or gated with `onlyKeepers`. Just make it `internal`: https://github.com/hsukev/KeeperDAO/blob/a2b65aee342e5217678e9a3a8649ba5f82dfa9a8/contracts/Strategy.sol#L201

**Comment by matkam** [comment]
Is it worth calling `harvest()` here after the `claim(...)`? https://github.com/hsukev/KeeperDAO/blob/a2b65aee342e5217678e9a3a8649ba5f82dfa9a8/contracts/Strategy.sol#L207

**Comment by hsukev** [comment]
> Is there any other way to test these? This is not a good idea to leave publicly accessible.
> https://github.com/hsukev/KeeperDAO/blob/a2b65aee342e5217678e9a3a8649ba5f82dfa9a8/contracts/Strategy.sol#L90
> https://github.com/hsukev/KeeperDAO/blob/a2b65aee342e5217678e9a3a8649ba5f82dfa9a8/contracts/Strategy.sol#L211
@matkam 
Oops, those were added for myself for testing when interactive broke. Same with the `netPositive` function. Let me change them.

**Comment by hsukev** [comment]
> So `harvest()` needs to be called separately? Is there a way to harvest (or `report(...)` and `vault.adjustPosition(...)`) so you don't need a separate call? https://github.com/hsukev/KeeperDAO/blob/a2b65aee342e5217678e9a3a8649ba5f82dfa9a8/contracts/Strategy.sol#L207

I think you might be asking about the strategy's `claimReward` instead of harvest()? Reward claim can't really be incorporated into `harvest()` bc it needs off-chain data. So it has to be manually called periodically

**Comment by hsukev** [comment]
added a single function that claims and harvests

**Comment by TheRyanMiller** [comment]
You have two unnecessary approvals:
- [one](https://github.com/hsukev/KeeperDAO/blob/c42ed390b99cb6b33c47ce3927d6b53f9eaaed78/contracts/Strategy.sol#L40)
- [two](https://github.com/hsukev/KeeperDAO/blob/c42ed390b99cb6b33c47ce3927d6b53f9eaaed78/contracts/Strategy.sol#L42)

**Comment by TheRyanMiller** [comment]
In your `prepareReturn` you should have an `else` condition for [this `if` condition](https://github.com/hsukev/KeeperDAO/blob/c42ed390b99cb6b33c47ce3927d6b53f9eaaed78/contracts/Strategy.sol#L104) which accounts for when your `incurredLoss > profit`. 
In this scenario you will want to record a loss of `_loss = incurredLosses.sub(profit)`.

**Comment by TheRyanMiller** [comment]
In [`prepareReturn`](https://github.com/hsukev/KeeperDAO/blob/c42ed390b99cb6b33c47ce3927d6b53f9eaaed78/contracts/Strategy.sol#L90) you should consider just keeping a running total of profits using the already exiting `_profit` variable.
Doing this will save you some gas. I don't think I see a need for you to have both `profit` and `usableProfit` variables.

**Comment by TheRyanMiller** [comment]
Don't do this.
This will cost the user more money on withdraw for no added benefit.

https://github.com/hsukev/KeeperDAO/blob/c42ed390b99cb6b33c47ce3927d6b53f9eaaed78/contracts/Strategy.sol#L152-L155

**Comment by TheRyanMiller** [comment]
Why not save value of `balanceOfUnstaked()` right before this as you will use same value again in the next line. Save some gas.

https://github.com/hsukev/KeeperDAO/blob/c42ed390b99cb6b33c47ce3927d6b53f9eaaed78/contracts/Strategy.sol#L158

**Comment by TheRyanMiller** [comment]
want to make sure you know that `harvest()` is implemented in `BaseStrategy`. You have your own custom implementation which should not be necessary.

I think that `claimRewards()` is the special fxn which requires off-chain data. I recommend making this a dedicated function accessible to strategist/keeper/gov only. And within that function call `harvest()`.

**Comment by hsukev** [comment]
> For your harvest:
> I think you might not need to fully reimplement from BaseStrategy as you have done here.
> 
> My understanding is that `claimRewards()` is your "special" fxn which requires off-chain data. I'd recommend making that function external and accessible only to strategist/keeper/gov. Within that function, make a call to `super.harvest()`.
> 
> Let me know if you agree. But I think you can get rid of your custom harvest implementation and do something like this...
> 
> ```
> claimRewards(address _to, uint256 _earningsToDate, uint256 _nonce, bytes memory _signature, bool claimFirst) external onlyKeepers {
>    if (claimFirst) {
>             claimRewards(_to, _earningsToDate, _nonce, _signature);
>             super.harvest();
>     }
> }
> ```

unfortunately `harvest` is `external` ): that's why in this strategy, the custom harvest had to copy everything from the `BaseStrategy`

**Comment by hsukev** [comment]
> > For your harvest:
> > I think you might not need to fully reimplement from BaseStrategy as you have done here.
> > My understanding is that `claimRewards()` is your "special" fxn which requires off-chain data. I'd recommend making that function external and accessible only to strategist/keeper/gov. Within that function, make a call to `super.harvest()`.
> > Let me know if you agree. But I think you can get rid of your custom harvest implementation and do something like this...
> > ```
> > claimRewards(address _to, uint256 _earningsToDate, uint256 _nonce, bytes memory _signature, bool claimFirst) external onlyKeepers {
> >    if (claimFirst) {
> >             claimRewards(_to, _earningsToDate, _nonce, _signature);
> >             super.harvest();
> >     }
> > }
> > ```
> 
> unfortunately `harvest` is `external` ): that's why in this strategy, the custom harvest had to copy everything from the `BaseStrategy`

Further discussion with @TheRyanMiller led to the decision to remove the custom harvest function and opt for a simpler flow for harvesting rewards:
1. `claimReward()` manually by `gov`
2. `harvestTrigger()` by `keeper`
3. `harvest()` by `keeper`

**Comment by matkam** [comment]
> 1. `claimReward()` manually by `gov`

It'll be the strategist calling this instead of gov. You can use `onlyAuthorized`.
